### PR TITLE
feat(gitfs): go-git backed implementation of fs.FS

### PR DESCRIPTION
**Pull request by GeorgeMac**
Supports FLI-288

This adds the new `gitfs` internal package.
The package converts a go-git `Tree` instance into an implementation of `fs.FS`.
It includes a number of helper function to go from a `git.Repository` directly into an `fs.FS` based on a desired reference.

This will be used in the fs.GitStore implementation in a subsequent PR.

### CODE REVIEW

File: internal/gitfs/gitfs.go [hunk 1]
```diff
@@ -0,0 +1,349 @@
...
+}
+
+// WithReference overrides the default reference to main.
+func WithReference(ref plumbing.ReferenceName) containers.Option[Options] {
+	return func(o *Options) {
+		o.ref = ref
+	}
+}
+
+// NewFromRepo is a convience utility which constructs an isntance of FS
```
Comment (yquansah) [inline internal/gitfs/gitfs.go:47]:
```suggestion
// NewFromRepo is a convenience utility which constructs an instance of FS
```

File: internal/gitfs/gitfs.go [hunk 2]
```diff
@@ -0,0 +1,349 @@
...
+
+	ref, err := repo.Reference(o.ref, true)
+	if err != nil {
+		return FS{}, fmt.Errorf("resolving reference (%q): %w", o.ref, err)
+	}
+
+	return NewFromRepoHash(repo, ref.Hash())
+}
+
+// NewFromRepoHash is a convience utility which constructs an isntance of FS
```
Comment (yquansah) [inline internal/gitfs/gitfs.go:63]:
```suggestion
// NewFromRepoHash is a convenience utility which constructs an instance of FS
```

File: internal/gitfs/gitfs.go [hunk 3]
```diff
@@ -0,0 +1,349 @@
...
+func (f FS) Open(name string) (_ fs.File, err error) {
+	defer func() {
+		if err != nil {
+			err = opPathError("Open", name)(err)
+		}
+	}()
+
+	if !fs.ValidPath(name) {
+		return nil, fs.ErrInvalid
+	}
```
Comment (yquansah) [inline internal/gitfs/gitfs.go:96]: Just clarifying my understanding here, this picks up the deferred error wrapping because of the named return values?

Comment (GeorgeMac) [reply internal/gitfs/gitfs.go:96]: Thats correct 👍 The defer runs after `return` at which point `err` has been populated by whatever the `return` statement has in that argument position.

This is a lazy way of wrapping all things 😂 :
https://go.dev/play/p/lG4C8IUiW7C

Comment (yquansah) [reply internal/gitfs/gitfs.go:96]: @GeorgeMac Makes sense. Thanks!

File: internal/gitfs/gitfs.go [hunk 4]
```diff
@@ -0,0 +1,350 @@
...
+
+	entry := &object.TreeEntry{
+		Name: ".",
+		Mode: filemode.Dir,
+		Hash: f.tree.Hash,
+	}
+
+	if name != "." {
```
Comment (yquansah) [inline internal/gitfs/gitfs.go:104]: What would the `f.tree.FindEntry(".")` return? It seems we are guarding against the case to not feed a period into that function (just curious)

**Review by yquansah — APPROVED** [review]
This looks great to me. QQ on directories here, would there be a case where we would be passing an actual directory to `gitfs.Open(...)`? Is the `DirEntry` functionality needed because of `fs.WalkDir`?

File: internal/gitfs/gitfs.go (hunk 4 above)
Comment (GeorgeMac) [reply internal/gitfs/gitfs.go:104]: Good question. Maybe this needs a comment.

When I first spiked this I found that the Git tree does not have an entry for `"."`.
Git isn't really a Posix filesystem, and relative links like this don't appear to exist.

In Git there are a few types of objects, but mostly beneath a `commit` object there are `trees` and `blobs`.
- `trees` == `dirs`
- `blobs` == `file contents`

A `commit` points to a `tree` object. This is the root of the tree and can be considered the root directory.

Long story short, there is no path `"."` in the git Tree object. But I fabricate one here to seed the traversal.

**Comment by GeorgeMac** [comment]
> This looks great to me. QQ on directories here, would there be a case where we would be passing an actual directory to gitfs.Open(...)? Is the DirEntry functionality needed because of fs.WalkDir?

@yquansah that is correct. Well more specifically, it is needed for a complete implementation of `fs.FS`.

`Open` returns an instance of https://pkg.go.dev/io/fs#File (which is a minimal interface for any file or dir).

If the path passed to `Open` refers to a directory then you need to return an implementation of: https://pkg.go.dev/io/fs#ReadDirFile
This is where the `ReadDir` method needs to additionally be added.

You're right that `fs.Walk` leverages `ReadDir` on directory file types to descend into the entire tree.

File: internal/gitfs/gitfs.go (hunk 4 above)
Comment (yquansah) [reply internal/gitfs/gitfs.go:104]: Right! No POSIX semantics that makes sense.

**Review by yquansah — APPROVED** [review]
I appreciate the tests here! Helps me understand better.
